src/rust/ide/src/double_representation/graph.rs

```diff
@@ -66,28 +95,66 @@ pub fn traverse_for_definition
 /// Description of the graph, based on information available in AST.
 #[derive(Clone,Debug)]
 pub struct GraphInfo {
-    source : DefinitionInfo,
-    /// Describes all known nodes in this graph (does not include special pseudo-nodes like graph
-    /// inputs and outputs).
-    pub nodes : Vec<node::NodeInfo>,
+    source:DefinitionInfo,
 }
 
 impl GraphInfo {
     /// Describe graph of the given definition.
     pub fn from_definition(source:DefinitionInfo) -> GraphInfo {
-        let nodes = Self::from_function_binding(source.ast.clone());
-        GraphInfo {source,nodes}
+        GraphInfo {source}
     }
 
     /// Lists nodes in the given binding's ast (infix expression).
-    fn from_function_binding(ast:known::Infix) -> Vec<node::NodeInfo> {
+    fn from_function_binding(ast:known::Infix) -> Vec<NodeInfo> {
         let body = ast.rarg.clone();
         if let Ok(body_block) = known::Block::try_new(body.clone()) {
             block_nodes(&body_block)
         } else {
             expression_node(body)
         }
     }
+
+    /// Gets all known nodes in this graph (does not include special pseudo-nodes like graph
+    /// inputs and outputs).
+    pub fn nodes(&self) -> Vec<NodeInfo> {
+        Self::from_function_binding(self.source.ast.clone())
+    }
+
+    /// Adds a new node to this graph.
+    pub fn add_node
+    (&mut self, line_ast:Ast, location_hint:LocationHint) -> FallibleResult<()> {
+        let mut lines = self.source.block_lines()?;
+
+        let find_position = |id| {
+            lines.iter().find_position(|line| {
+                line.elem.as_ref().map(|line_ast| {
+                    NodeInfo::from_line_ast(line_ast).map(|node| node.id() == id).unwrap_or(false)
+                }).unwrap_or(false)
+            }).map(|(index,_)| index).ok_or(IDNotFound{id})
+        };
```

Comment:
This is really complex. Please refactor to variables and smaller expressions. Variables should be named in such way that we can understand what is hapenning on each step.

Reply:
+1